contracts/account/utils/draft-ERC4337Utils.sol

```diff
@@ -36,13 +36,33 @@ library ERC4337Utils {
     /// @dev For simulation purposes, validateUserOp (and validatePaymasterUserOp) must return this value in case of signature failure, instead of revert.
     uint256 internal constant SIG_VALIDATION_FAILED = 1;
 
-    /// @dev Parses the validation data into its components. See {packValidationData}.
+    /// @dev Magic value used in EntryPoint v0.9+ to detect the presence of a paymaster signature in `paymasterAndData`.
+    bytes8 internal constant PAYMASTER_SIG_MAGIC = 0x22e325a297439656;
+
+    /// @dev Highest bit set to 1 in a 6-bytes field.
+    uint48 internal constant BLOCK_RANGE_MASK = 0x800000000000;
+
+    /// @dev Validity range of the validation data.
+    enum ValidationRange {
+        TIMESTAMP,
+        BLOCK
+    }
+
+    /// @dev Parses the validation data into its components and the validity range. See {packValidationData}.
     function parseValidationData(
         uint256 validationData
-    ) internal pure returns (address aggregator, uint48 validAfter, uint48 validUntil) {
+    ) internal pure returns (address aggregator, uint48 validAfter, uint48 validUntil, ValidationRange range) {
         validAfter = uint48(bytes32(validationData).extract_32_6(0));
         validUntil = uint48(bytes32(validationData).extract_32_6(6));
         aggregator = address(bytes32(validationData).extract_32_20(12));
+        range = ValidationRange.TIMESTAMP;
+
+        if (validAfter >= BLOCK_RANGE_MASK && validUntil >= BLOCK_RANGE_MASK) {
```

Comment:
what if only one of the two has the flag. Is it ok to pass it like that, and handle it as a timestamp ?

Reply:
> Is it ok to pass it like that, and handle it as a timestamp ?

Yes, it is what the entrypoint does, so even if one of the values has the flag set, it will be ignored by the entrypoint